### PR TITLE
Multiple bundle config

### PR DESCRIPTION
**Pull request by howbazaar**
When deploying complex models, often the bundle is broken up into multiple components.

This branch also fixes the problem with YAML anchors used on the bundle config.

## QA steps

Create two files:
remove-easyrsa.yaml (http://pastebin.ubuntu.com/25929417/)
placement.yaml (http://pastebin.ubuntu.com/25929422/)
Then run the following in an empty model:
$ juju deploy canonical-kubernetes --bundle-config ~/canonical/remove-easyrsa.yaml --bundle-config ~/canonical/placement.yaml --dry-run

The output should match http://pastebin.ubuntu.com/25929429/ where there is no easyrsa deployed, and all the units are colocated on three machines.

## Documentation changes

The bundle documentation will need to be updated to outline the precedence rules for the config with respect to the machines, applications, and series. It should also explain how to use config to remove an application from a bundle, as well as how to use YAML anchors for common config.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1721603
https://bugs.launchpad.net/juju/+bug/1728593

### CODE REVIEW

File: cmd/juju/application/bundle.go [hunk 1]
```diff
@@ -1105,44 +1105,54 @@ func processValue(baseDir string, v interface{}) (interface{}, bool, error) {
...
-		return errors.Annotate(err, "unable to normalise bundle-config file")
-	}
-	// Make sure the filename is absolute.
-	if !filepath.IsAbs(bundleConfigFile) {
-		cwd, err := os.Getwd()
+func processBundleConfig(data *charm.BundleData, bundleConfigFiles ...string) error {
+	for _, filename := range bundleConfigFiles {
+		bundleConfigFile, err := utils.NormalizePath(filename)
 		if err != nil {
+			return errors.Annotate(err, "unable to normalise bundle-config file")
```
Comment (anastasiamac) [inline cmd/juju/application/bundle.go]: Could we please add a filename into the error message?

File: cmd/juju/application/bundle.go [hunk 2]
```diff
@@ -1105,44 +1105,54 @@ func processValue(baseDir string, v interface{}) (interface{}, bool, error) {
...
+		if err := processSingleBundleConfig(data, bundleConfigFile); err != nil {
 			return errors.Trace(err)
 		}
-		bundleConfigFile = filepath.Clean(filepath.Join(cwd, bundleConfigFile))
 	}
+	return nil
+}
+
+func processSingleBundleConfig(data *charm.BundleData, bundleConfigFile string) error {
+
```
Comment (anastasiamac) [inline cmd/juju/application/bundle.go]: Not sure that empty line is helping readability...

File: cmd/juju/application/bundle.go [hunk 3]
```diff
@@ -1105,44 +1105,54 @@ func processValue(baseDir string, v interface{}) (interface{}, bool, error) {
...
+	}
+
+	// From here we walk through the new bundleData, and override values
+	// in the current bundle.
+
+	// If a new application is added, the content is added.
+	// If an application is defined but with no content, that means
+	// remove the application from the parent bundle.
+	//   - if this is happening, all related relations are also removed.
+	// If the application exists in both, values here override values there.
```
Comment (anastasiamac) [inline cmd/juju/application/bundle.go]: This file precedence - order matters - should definitely be documented in either help or online docs too :D

File: cmd/juju/application/deploy.go [hunk 4]
```diff
@@ -264,9 +264,9 @@ type DeployCommand struct {
 	// or a bundle name.
 	CharmOrBundle string
 
-	// BundleConfig refers to a config file that specifies application overrides, and
-	// in the near future, machine and space mappings.
-	BundleConfigFile string
+	// BundleConfig refers to config files that specifies additional
```
Comment (anastasiamac) [inline cmd/juju/application/deploy.go]: typo - '..files that specify.."

File: cmd/juju/application/bundle.go (hunk 1 above)
Comment (howbazaar) [reply cmd/juju/application/bundle.go]: The filename is already in the text from NormalizePath.

File: cmd/juju/application/bundle.go (hunk 2 above)
Comment (howbazaar) [reply cmd/juju/application/bundle.go]: Removed.

File: cmd/juju/application/bundle.go (hunk 3 above)
Comment (howbazaar) [reply cmd/juju/application/bundle.go]: Yes.

File: cmd/juju/application/deploy.go (hunk 4 above)
Comment (howbazaar) [reply cmd/juju/application/deploy.go]: Fixed.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**